### PR TITLE
[Feature] Add JSONPath response filtering

### LINKED ISSUE #905: [Feature Request] JSON path filtering on response body
Add a text input next to the response body which accepts a JSON path to filter the data

JSON path : https://goessner.net/articles/JsonPath/

A JS library for JSON path : https://www.npmjs.com/package/jsonpath-plus

### PR DESCRIPTION
**Pull request by Oryss**
# Description

Adds an input at the bottom of the response where you can enter JSONPath expressions (https://goessner.net/articles/JsonPath/)

Example URL : https://jsonplaceholder.typicode.com/posts

Example JSONPath expression : 
- `$..[?(@.userId == 5)]` :  Display posts where userId is 5
- `$..title` : Display all titles

image

By changing a few conditions and adding an Xpath library we could handle xpath filtering the same way

resolves #905  

### Contribution Checklist:

- [x] **The pull request only addresses one issue or adds one feature.**
- [x] **The pull request does not introduce any breaking changes**
- [x] **I have added screenshots or gifs to help explain the change if applicable.**
- [x] **I have read the [contribution guidelines](https://github.com/usebruno/bruno/blob/main/contributing.md).**
- [x] **Create an issue and link to the pull request.**

Note: Keeping the PR small and focused helps make it easier to review and merge. If you have multiple changes you want to make, please consider submitting them as separate pull requests.

### Publishing to New Package Managers

Please see [here](../publishing.md) for more information.

### CODE REVIEW

File: packages/bruno-app/src/components/ResponsePane/QueryResult/StyledWrapper.js [hunk 1]
```diff
@@ -3,7 +3,8 @@ import styled from 'styled-components';
 const StyledWrapper = styled.div`
   display: grid;
   grid-template-columns: 100%;
-  grid-template-rows: 1.25rem calc(100% - 1.25rem);
+  grid-template-rows: ${(props) =>
+    props.queryFilterEnabled ? '1.25rem calc(100% - 3.5rem)' : '1.25rem calc(100% - 1.25rem)'};
```
Comment (Its-treason) [inline packages/bruno-app/src/components/ResponsePane/QueryResult/StyledWrapper.js]:
```suggestion
    props.queryFilterEnabled ? '1.25rem 1fr 2.25rem' : '1.25rem 1fr'};
```
You can utilize the CSS-Grid here and add grid row for your input. I also suggest to change `calc(100% - 1.25rem)` to `1fr` to make it cleaner.

File: packages/bruno-app/src/components/ResponsePane/QueryResult/QueryResultFilter/index.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  return (
+    <div className={'response-filter relative'}>
+      <div className="absolute inset-y-0 left-0 pl-4 flex items-center">
+        <div className="text-gray-500 sm:text-sm" id="request-filter-icon">
+          <IconFilter size={16} strokeWidth={1.5} />
+        </div>
+      </div>
+
+      {tooltipText && typeof tooltipText === 'string' && (
```
Comment (Its-treason) [inline packages/bruno-app/src/components/ResponsePane/QueryResult/QueryResultFilter/index.js]: You don't need the null & typeof check here. If `tooltipText` is null, the tooltip will not be shown.

**Review by Its-treason — CHANGES_REQUESTED** [review]
Looks very good! I found 2 small things, but your implementation looks very good overall! Looking forward to having this feature in the future.

File: packages/bruno-app/src/components/ResponsePane/QueryResult/QueryResultFilter/index.js (hunk 2 above)
Comment (Oryss) [reply packages/bruno-app/src/components/ResponsePane/QueryResult/QueryResultFilter/index.js]: It's to hint the type to the IDE and be safe but I could also add   /** @type {string} */ to the useMemo line 6

image

**Review by Its-treason — APPROVED** [review]

**Comment by helloanoop** [comment]
This is great! Thank you @Oryss !

As always, thanks for the review @Its-treason !

**Comment by bbbco** [comment]
Why did you choose JSON path and not the more powerful [JSONata](https://jsonata.org/)? It's also [backwards compatible with JSON Path](https://try.jsonata.org/hcW-iQtDG).

**Comment by Oryss** [comment]
> Why did you choose JSON path and not the more powerful [JSONata](https://jsonata.org/)? It's also [backwards compatible with JSON Path](https://try.jsonata.org/hcW-iQtDG).

Because I didn't know about JSONata until I read your comment :p I'll take a look at it, but the filtering could easily be swapped and extended to other structures like XML (xpath)

**Comment by bw-hro** [comment]
Very nice feature. Would it maybe possible to remember the provided filter when switching between requests? Also the possibility to toggle application of filter by toggling the filter button would be nice.

**Comment by hungdh0x5e** [comment]
@Oryss can you add [jq](https://github.com/jqlang/jq) ?

**Pull request merged**